### PR TITLE
Branch Deletion

### LINKED ISSUE #388: Deleting branches does not work
**Describe the bug**

`branch.delete()` does nothing

**To Reproduce**

Code for https://codepen.io/nicoespeon/pen/arqPWb?editors=0010

```
const graphContainer = document.getElementById("graph-container");
 
// Instantiate the graph.
const gitgraph = GitgraphJS.createGitgraph(graphContainer);

// Simulate git commands with Gitgraph API.
const master = gitgraph.branch("master");
master.commit("Initial commit");

var feature = master.branch("feature");
feature.commit("a");
master.merge(feature);
feature.delete();
```
Branch "feature" is still visible.

**Expected behavior**

The ref "feature" should be gone.

**Screenshots**

n/a

**Context**

`@gitgraph/js`

### PR DESCRIPTION
**Pull request by acransac**
This patch adds the possibility to delete a branch. It extends the branch user API so one can call `someBranch.delete()` and see the branch disappear from the rendered graph if it is not merged before.

Concretely, it clears out the branch name from the graph's `branches` property and from its `refs`. Commits and tags from the branch remain in the graph. Once a branch is deleted, the branch user API throws for the operations that don't make sense anymore. It is also impossible to delete a branch when it is checked out.

On the rendering side, only commits that are not reachable somehow disappear from view. This means that if the deleted branch was merged beforehand, the commits stay visible.

To the best of my understanding of the code, this change is compatible with previous use and behavior (notably, all tests pass without update). I also added tests and stories to cover what I added.

Closes #388

### CODE REVIEW

File: packages/stories/src/gitgraph-js/1-basic-usage.stories.tsx [hunk 1]
```diff
@@ -389,4 +389,37 @@ storiesOf("gitgraph-js/1. Basic usage", module)
...
+        const master = gitgraph.branch("master").commit("Initial commit");
+        const develop = gitgraph.branch("develop");
+        develop.commit("one");
+        master.commit("two");
+        master.checkout();
+        develop.delete();
+      }}
+    </GraphContainer>
+  ))
+  .add("delete a branch after merging it", () => (
```
Comment (nicoespeon) [inline packages/stories/src/gitgraph-js/1-basic-usage.stories.tsx:409]: 👍 Perfect, thanks for adding stories to this!

File: packages/gitgraph-core/src/user-api/branch-user-api.ts [hunk 2]
```diff
@@ -119,6 +185,10 @@ class BranchUserApi<TNode> {
    */
   public merge(options: GitgraphMergeOptions<TNode>): this;
   public merge(...args: any[]): this {
+    if (this._branch.isDeleted() && !this._isReferenced()) {
+      throw new Error(`Cannot merge to the deleted branch "${this.name}"`);
```
Comment (nicoespeon) [inline packages/gitgraph-core/src/user-api/branch-user-api.ts:189]: 👌 That makes sense and is a useful error to debug.

File: packages/gitgraph-core/src/user-api/branch-user-api.ts [hunk 3]
```diff
@@ -98,6 +106,64 @@ class BranchUserApi<TNode> {
...
+   */
+  public delete(): this {
+    // Delete all references to the branch from the graph (graph.branches and graph.refs)
+    // and from the commits (commit.refs). Then, make the branch instance a deleted branch.
+    // Like in git, the commits and tags in the deleted branch remain in the graph
+    if (
+      this._graph.refs.getCommit("HEAD") ===
+      this._graph.refs.getCommit(this.name)
+    ) {
+      throw new Error(`Cannot delete the checked out branch "${this.name}"`);
```
Comment (nicoespeon) [inline packages/gitgraph-core/src/user-api/branch-user-api.ts:120]: 👍 Nice, you thought about this edge-case

File: packages/gitgraph-core/src/__tests__/gitgraph.getRenderedData.branches.test.ts [hunk 4]
```diff
@@ -221,4 +221,54 @@ describe("Gitgraph.getRenderedData.branches", () => {
       ]);
     });
   });
+
+  describe("deleted branch", () => {
+    let gitgraph;
```
Comment (nicoespeon) [inline packages/gitgraph-core/src/__tests__/gitgraph.getRenderedData.branches.test.ts:226]: 🍰 We usually prefer to declare variables closer to their usage, directly into each test. It makes it easier to ensure each test is decoupled from the other (no shared variable).

It's OK here 'cause there are only 2 tests, but if you can refactor this it will be great!

Note that if we want to avoid repeating the same code statements to init the graph state, we can still extract them in a function instead, like so:

```ts
    it("should ignore commits on a deleted branch", () => {
      const { gitgraph, develop } = createGraphWithSomeCommits();

      develop.delete();

      const { commits } = gitgraph._graph.getRenderedData();

      expect(commits).toMatchObject([
        { subject: "one", refs: [] },
        { subject: "two", refs: ["master", "HEAD"] },
      ]);
    });
```

File: packages/gitgraph-core/src/__tests__/branch.test.ts [hunk 5]
```diff
@@ -188,4 +188,126 @@ describe("Branch", () => {
       });
     });
   });
+
+  describe("delete", () => {
+    let gitgraph;
```
Comment (nicoespeon) [inline packages/gitgraph-core/src/__tests__/branch.test.ts:193]: 🍰 We usually prefer to declare variables closer to their usage, directly into each test. It makes it easier to ensure each test is decoupled from the other (no shared variable).

Note that if we want to avoid repeating the same code statements to init the graph state, we can still extract them in a function instead, like so:

```ts
    it("should be deleted and not referenced", () => {
      const { develop, feature } = createGraphWithSomeCommits();

      develop.checkout();

      feature.delete();

      expect(feature._branch.isDeleted() && !feature._isReferenced()).toBe(
        true,
      );
    });
```

**Review by nicoespeon — APPROVED** [review]
This PR is 💯  

- It has a lot of tests to prove that it's working—honestly, it also makes it easier to review as I can start reading the tests
- It has useful comments to explain "why" we do this
- It doesn't break other tests & stories, which means no existing behavior was changed

So 👏  I'm gonna ship this with the next release.

**Comment by nicoespeon** [comment]
@all-contributors please add acransac for code, bug

File: packages/gitgraph-core/src/__tests__/branch.test.ts (hunk 5 above)
Comment (acransac) [reply packages/gitgraph-core/src/__tests__/branch.test.ts:193]: Hi @nicoespeon, thanks for the quick review and kind feedback! Should I make a PR to refactor these tests?
